### DISCUSSION TITLE
SNMP protocol support

### DISCUSSION THREAD

**Discussion by hensephil** [arendst/Tasmota#16593] [Ideas]
I would love it if Tasmota had SNMP support. With this I could also query certain values ​​via my monitoring system (how are there updates, what is the IP, which WLAN is the device connected to, ...)

Regards
Philipp

**Comment by sfromis** [comment]
In general, features gets added by someone who would love to have them, and do the coding to get it working within the structure of Tasmota, in a good way, and submit a PR. Or maybe do a gateway using existing Tasmota features for the device info, and interface to SNMP monitoring.

**Comment by s-hadinger** [comment]
There is very low chance that SNMP is added in the code base, considering that SNMP is not common in hobbyists houses.

If you are courageous enough, it should be possible in Berry to develop an SNMP client. Berry already supports UDP.

**Comment by nixator** [comment]
I tryed to add SNMP into Tasmota and it worked but it was very dirty experiment but it worked... I needed it to monitor energymeter via Zabbix. Finaly I switched to HTTP API, because Zabbix can read data from tasmota via HTTP request... https://github.com/0neblock/Arduino_SNMP

**Reply by hensephil** [reply]
Hello,

Do you have instructions on how I can embed this?

**Reply by xtvdata** [reply]
No need to use insecure non-TLS APIs, nor to implement SNMP.
Zabbix can receive data directly from MQTT ;-)

**Reply by nixator** [reply]
exactly, I run zabbix with MQTT client working. Only caveat is that you must add all hosts to config file of agent... But it work's

**Comment by Npeca75** [comment]
> I tryed to add SNMP into Tasmota and it worked but it was very dirty experiment but it worked... 
Hi


is there any chance to write some kind of step-by-step tutorial ?
my whole network monitoring is based on SNMP/LibreNMS and i wanted to incorporate dozen of sonoff device on remote locations ... so i need this SNMP feature badly
please?

**Reply by barbudor** [reply]
SNMP will not be added to Tasmota in an official way.
The standard protocol for IOT and Home Automation is MQTT.
You'd better look at building a gateway between MQTT and SNMP

**Reply by Npeca75** [reply]
@barbudor 
>SNMP will not be added to Tasmota in an official way.

ok, my point of view was:
plastic, cheap wifi routers (10-15 eur) are extended with OpenWRT -> SNMP/LLDP/FDB to act as semi professional devices
so why not do the same with plastic, cheap sonoff (10-15 eur) ?
OpenWRT is alternative opensource firmware
Tasmota is alternative opensource firmware

ok, so, the hard way, get the DEV ENV and start to thinkering

**Reply by s-hadinger** [reply]
Just curious, why do you need snmp? For querying value? If yes what values? For alerts?

**Comment by Npeca75** [comment]
@s-hadinger 
> Just curious, why do you need snmp? 

how to explain ? it is a kind of "life style"
network equipment are grouped in LibreNMS. 
All IPs, routes, FDBs, VLANs, MACs, all temperature/humidity/freq/fan speeds ...etc
it is mandatory to have all these information in one place with 300+ USER devices on 60-70 network devices + cca 10 servers

so, my idea is, that sonoff IS network device, providing useful data
Power consumption, Voltage, Temperature, Humidity ...
look at this like LMSensors, or like UPS-NUT
they all have place in network environment, and yes, they provide SNMP data

sonoff could become part of this "remote-sensor-with-snmp" group in semi prof setups
let's say you place sonoff in RACK to monitor voltage/temp/power/humidity
and whole RACK is on SNMP, only sonoff is ... something else :(

i hope this will bring SNMP support forward

**Reply by sfromis** [reply]
To complicate things a bit....

Expecting that SNMP is common in enterprise environments, but not in small-scale setups like homes, I'm wondering about how much overlap there would be in a Venn diagram of enterprises where strict certification/infrastructure requirements for equipment makes devices "hacked" with open source firmware a no-go (unless paying a high certification cost), small-scale users who do not worry much about certification, and devices even available to such hacking.

Unsurprisingly, my expectation would be not much overlap....

**Reply by Npeca75** [reply]
@sfromis 
>but not in small-scale setups like homes

that is the whole point of story
why limit sonoff to home/hobby ?


when TP-LINK or DLINK or other cheap plastic manufacturer introduced they APs, opensource community bring them to semi prof. use with alternative firmware
yes, one could ask, is 15 eur device have a place in enterprise? no, maybe not, but they have a place in small office/small ISP and similar mid-sized networks. It is a good HW with s***y SW
i will like it very much if Tasmota bring this (small) revolution similar like OpenWRT does with with cheap devices

**Reply by sfromis** [reply]
> why limit sonoff to home/hobby

I already explained that. Strict certification requirements for the full device (both hardware and firmware) are quite often seen in enterprise environments. And the high price of getting that is not very compatible with open source projects and contributors.

**Reply by s-hadinger** [reply]
Anyways, remapping all Tasmota sensors to SNMP OIDs is a huge task that is unlikely to happen. Tasmota can easily provide JSON but it's of no use for SNMP. 

You could probably handcraft some simple SNMP UDP packets of you have a specific need. But don't expect Tasmota to provide full SNMP support anytime soon, if ever

**Reply by xtvdata** [reply]
Moreover, SNMP only makes sense (in business environment, but if you ask me also in home applications) if it's used with a secured v3 protocol; v1/v2 protocols could be implemented (maybe via Berry), however v3 is not so simple (at least for Tasmota devices).

As mentioned above, if you need a monitoring solution that can receive data via MQTT natively, you could consider Zabbix.

P.S.: if I'd have to consume all those resources on a Tasmota device, for sure I'd go for other features (e.g.: upgrade web interface/API to use TLS, maybe pulling certificates from an ACME private server, unfortunately ACME is probably too advanced for home applications...)

**Reply by Npeca75** [reply]
@xtvdata 
>you could consider Zabbix

sorry, our monitoring infrastructure is based on LNMS and distributed pollers
furthermore, i find LNMS more superior in some parts, like LLDP and drawing network infrastructure

so there is 2 possible way
to extend LNMS with API client to read API/JSON from Tasmoa or get my hands dirty and start to modify Tasmota ... :(


@s-hadinger 
>SNMP support anytime soon, if ever

yes, i understand this is a huge job
it was idea to make Tasmoa more enterprise friendly 
hope it will be implemented one day

**Reply by barbudor** [reply]
I would expect a proper NMS to be able to deal with more protocol than just SNMP
If not, a gateway from MQTT to SNMP is probably the best solution

**Reply by xtvdata** [reply]
Anyway, today any unencrypted traffic, even in a local network, is considered insecure and unfit for a business environment.
For such an environment, I wouldn't recommend use any device/firmware/software provided without certification and a very active security patch channel (E.g.: that's why only a bunch of linux distributions are considered usable in an enterprise context).

However, if you really want to use Tasmota, your only reasonably safe options are:
- MQTT with TLS support,
- write a C++ (or Berry) extension to implement SNMPv3
- write a TLS endpoint in Berry (very complicated to maintain without a good automatic certificate distribution mechanism).
but, in any case, you should keep the standard http web admin/API disabled (as it's not encrypted).

P.S.: what would you want to use Tasmota devices for in an enterprise architecture/network? Even if it's just for monitoring, devices should be connected to a separate dedicated network behind a firewall (as all IoT should be)... and still I would consider it dangerous.

**Reply by Npeca75** [reply]
@xtvdata 
>Even if it's just for monitoring, devices should be connected to a separate dedicated network

thank you for your advice
yes, i am well aware about dedicated/protected mgmnt/monitoring (V)LAN 

> to implement SNMPv3

many older devices which are present in semi professional networks have only v2c, so yes, we are using many techniques to protect them on L2 and L3 levels

>what would you want to use Tasmota devices for in an enterprise architecture/network?

use case is environmental monitoring
sonoff devices are cheap and could be replaced at any time & any rate

why not use dedicated SNMP ENV Monitor?
price is only one of factors

mostly, it is availability
today, i will buy SNMP Monitor from manufacturer "AAA"
tomorrow, it will break, and manufacturer "AAA" no longer exist, so i need to buy from "BBB"

problem ? different sensors, different OIDs, ... start from scratch

this could be a big benefit of Tasmoa
every sonoff device will have for ex: powerConsumptionTable at OID x.y.z
in every release, on every HW
no more messing with different manufacturers, OIDs, templates and similar

**Reply by s-hadinger** [reply]
I understand, but like stated above, it would be much better to have MQTT -> SNMP gateway somewhere.

**Comment by Noschvie** [comment]
@hensephil : do you have a proposal for a MIB used for e.g. Shelly Plus 2PM ?

**Comment by Jibun-no-Kage** [comment]
WOULD LOVE TO HAVE SNMP SUPPORT!  Just add me to the YES for it side!  Have a long history with SNMP and it would be a benefit for sure!

**Reply by s-hadinger** [reply]
It's already answered. No developer from Tasmota will be doing SNMP support. Please use MQTT
